File: dashboard/src/api/ipam.ts
import { api } from "./client";

export type IpamAddressStatus = "Offline" | "Online" | "Reserved" | "Free" | "DHCP";

export interface IpamRegion {
  id: number;
  name: string;
  description: string | null;
  contact_name?: string | null;
  contact_email?: string | null;
  rack_notes?: string | null;
  internal_asn?: string | null;
  subnet_count?: number;
  address_count?: number;
}

export interface IpamRegionCreate {
  name: string;
  description?: string;
  contact_name?: string;
  contact_email?: string;
  rack_notes?: string;
  internal_asn?: string;
}

export interface IpamRegionUpdate {
  name?: string;
  description?: string | null;
  contact_name?: string | null;
  contact_email?: string | null;
  rack_notes?: string | null;
  internal_asn?: string | null;
}

export interface IpamSubnet {
  id: number;
  region_id: number;
  region_name: string | null;
  vlan_id: number | null;
  vlan_name: string | null;
  cidr_block: string;
  broadcast_domain: string | null;
  description: string | null;
  created_at: string | null;
  deleted_at?: string | null;
  scan_enabled?: boolean;
  scan_cron?: string | null;
  utilization_alert_pct?: number;
  utilization_webhook_url?: string | null;
  overlap_warnings?: { id: number; cidr_block: string }[];
  vlan_warnings?: { id: number; cidr_block: string; region_name: string }[];
  rfc1918_scope: string;
}

export interface IpamSubnetCreate {
  region_id: number;
  cidr_block: string;
  vlan_id?: number;
  vlan_name?: string;
  broadcast_domain?: string;
  description?: string;
  scan_enabled?: boolean;
  scan_cron?: string;
  utilization_alert_pct?: number;
  utilization_webhook_url?: string;
}

export interface IpamSubnetUpdate {
  region_id?: number;
  vlan_id?: number | null;
  vlan_name?: string | null;
  broadcast_domain?: string | null;
  description?: string | null;
  scan_enabled?: boolean;
  scan_cron?: string | null;
  utilization_alert_pct?: number;
  utilization_webhook_url?: string | null;
}

export interface IpamSubnetStatistics {
  subnet_id: number;
  cidr_block: string;
  region_id: number;
  vlan_id: number | null;
  total_host_capacity: number;
  occupied: number;
  free_tracked: number;
  free_remaining: number;
  utilization_percent: number;
  by_status: Record<string, number>;
  alert_threshold?: number;
  alert_triggered?: boolean;
}

export interface IpamAddress {
  id: number;
  subnet_id: number;
  ip_address: string;
  status: IpamAddressStatus;
  hostname: string | null;
  mac_address: string | null;
  description: string | null;
  last_seen: string | null;
  is_discovered_by_nmap: boolean;
  expires_at?: string | null;
  noc_device_id?: string | null;
  noc_hostname?: string | null;
  dhcp_lease_expires?: string | null;
  reservation_expired?: boolean;
  updated_at: string | null;
}

export interface IpamAddressPage {
  total: number;
  limit: number;
  offset: number;
  data: IpamAddress[];
}

export interface IpamAuditEntry {
  id: number;
  entity_type: string;
  entity_id: string;
  action: string;
  actor: string | null;
  changes: Record<string, unknown> | null;
  created_at: string;
}

export interface IpamHeatmap {
  subnet_id: number;
  cidr_block: string;
  prefixlen: number;
  cells: { ip: string; status: string; last_octet: number }[];
}

export interface IpamNmapDiscoverResult {
  subnet_id: number;
  cidr_block: string;
  hosts_capacity: number;
  hosts_up: number;
  created: number;
  updated: number;
  marked_offline: number;
  duration_ms: number;
  nmap_summary: string | null;
  noc_linked?: number;
}

export async function fetchIpamRegions() {
  const { data } = await api.get<IpamRegion[]>("/api/v1/ipam/regions");
  return data;
}

export async function createIpamRegion(body: IpamRegionCreate) {
  const { data } = await api.post<IpamRegion>("/api/v1/ipam/regions", body);
  return data;
}

export async function updateIpamRegion(id: number, body: IpamRegionUpdate) {
  const { data } = await api.patch<IpamRegion>(`/api/v1/ipam/regions/${id}`, body);
  return data;
}

export async function deleteIpamRegion(id: number) {
  await api.delete(`/api/v1/ipam/regions/${id}`);
}

export async function fetchIpamSubnets(regionId?: number) {
  const { data } = await api.get<IpamSubnet[]>("/api/v1/ipam/subnets", {
    params: regionId ? { region_id: regionId } : undefined,
  });
  return data;
}

export async function createIpamSubnet(body: IpamSubnetCreate) {
  const { data } = await api.post<IpamSubnet>("/api/v1/ipam/subnets", body);
  return data;
}

export async function updateIpamSubnet(id: number, body: IpamSubnetUpdate) {
  const { data } = await api.patch<IpamSubnet>(`/api/v1/ipam/subnets/${id}`, body);
  return data;
}

export async function deleteIpamSubnet(id: number) {
  await api.delete(`/api/v1/ipam/subnets/${id}`);
}

export async function fetchIpamSubnetStatistics(subnetId: number) {
  const { data } = await api.get<IpamSubnetStatistics>(`/api/v1/ipam/subnets/${subnetId}/statistics`);
  return data;
}

export async function fetchIpamSubnetAddresses(subnetId: number, page = 0, limit = 100) {
  const { data } = await api.get<IpamAddressPage>(`/api/v1/ipam/subnets/${subnetId}/addresses`, {
    params: { offset: page * limit, limit },
  });
  return data;
}

export async function searchIpam(q: string, offset = 0, limit = 50) {
  const { data } = await api.get<{ total: number; data: IpamAddress[] }>("/api/v1/ipam/search", {
    params: { q, offset, limit },
  });
  return data;
}

export async function fetchIpamHeatmap(subnetId: number) {
  const { data } = await api.get<IpamHeatmap>(`/api/v1/ipam/subnets/${subnetId}/heatmap`);
  return data;
}

export async function fetchIpamAudit(limit = 100) {
  const { data } = await api.get<IpamAuditEntry[]>("/api/v1/ipam/audit", { params: { limit } });
  return data;
}

export async function exportIpam(format: "json" | "csv", regionId?: number) {
  const { data } = await api.get("/api/v1/ipam/export", {
    params: { format, region_id: regionId },
    responseType: format === "csv" ? "text" : "json",
  });
  return data;
}

export async function importIpamSubnet(subnetId: number, rows: Record<string, unknown>[]) {
  const { data } = await api.post(`/api/v1/ipam/subnets/${subnetId}/import`, rows);
  return data as { created: number; updated: number; skipped: number };
}

export async function bulkReserveIpam(
  subnetId: number,
  body: { start_ip: string; end_ip: string; status?: IpamAddressStatus; description?: string; expires_at?: string },
) {
  const { data } = await api.post(`/api/v1/ipam/subnets/${subnetId}/addresses/bulk`, body);
  return data as { created: number; updated: number };
}

export async function syncDhcpIpam(subnetId: number, leases: Record<string, unknown>[]) {
  const { data } = await api.post(`/api/v1/ipam/subnets/${subnetId}/dhcp/sync`, leases);
  return data as { synced: number };
}

export async function linkNocAddress(addressId: number) {
  const { data } = await api.post(`/api/v1/ipam/addresses/${addressId}/link-noc`);
  return data;
}

export async function linkNocSubnet(subnetId: number) {
  const { data } = await api.post(`/api/v1/ipam/subnets/${subnetId}/link-noc-all`);
  return data as { linked: number };
}

export async function createIpamAddress(subnetId: number, body: Record<string, unknown>) {
  const { data } = await api.post<IpamAddress>(`/api/v1/ipam/subnets/${subnetId}/addresses`, body);
  return data;
}

export async function patchIpamAddress(addressId: number, body: Record<string, unknown>) {
  const { data } = await api.patch<IpamAddress>(`/api/v1/ipam/addresses/${addressId}`, body);
  return data;
}

export async function deleteIpamAddress(addressId: number) {
  await api.delete(`/api/v1/ipam/addresses/${addressId}`);
}

export async function discoverIpamSubnetNmap(subnetId: number, body = {}) {
  const { data } = await api.post<IpamNmapDiscoverResult>(
    `/api/v1/ipam/subnets/${subnetId}/discover`,
    { mark_offline: true, preserve_reserved: true, ...body },
    { timeout: 600_000 },
  );
  return data;
}

File: ipam/app/models/ipam.py
import enum

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.dialects.postgresql import CIDR, ENUM, INET, JSONB, MACADDR, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class IPAMAddressStatus(str, enum.Enum):
    OFFLINE = "Offline"
    ONLINE = "Online"
    RESERVED = "Reserved"
    FREE = "Free"
    DHCP = "DHCP"


ipam_status_pg = ENUM(
    IPAMAddressStatus,
    name="ipam_address_status",
    create_type=False,
    values_callable=lambda obj: [e.value for e in obj],
)


class IPAMRegion(Base):
    __tablename__ = "ipam_regions"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    contact_name: Mapped[str | None] = mapped_column(String(128))
    contact_email: Mapped[str | None] = mapped_column(String(255))
    rack_notes: Mapped[str | None] = mapped_column(Text)
    internal_asn: Mapped[str | None] = mapped_column(String(32))

    subnets: Mapped[list["IPAMSubnet"]] = relationship(back_populates="region")


class IPAMSubnet(Base):
    __tablename__ = "ipam_subnets"
    __table_args__ = (
        CheckConstraint(
            "vlan_id IS NULL OR (vlan_id >= 1 AND vlan_id <= 4094)",
            name="ipam_subnets_vlan_id_check",
        ),
        UniqueConstraint("region_id", "cidr_block", name="unique_cidr_per_region"),
        Index("idx_ipam_subnets_cidr_gist", "cidr_block", postgresql_using="gist"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    region_id: Mapped[int] = mapped_column(ForeignKey("ipam_regions.id", ondelete="RESTRICT"), nullable=False)
    vlan_id: Mapped[int | None] = mapped_column(Integer)
    vlan_name: Mapped[str | None] = mapped_column(String(64))
    cidr_block: Mapped[str] = mapped_column(CIDR, nullable=False)
    broadcast_domain: Mapped[str | None] = mapped_column(String(128))
    description: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[object] = mapped_column(DateTime(timezone=True), server_default=func.now())
    deleted_at: Mapped[object | None] = mapped_column(DateTime(timezone=True))
    scan_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    scan_cron: Mapped[str | None] = mapped_column(String(64))
    utilization_alert_pct: Mapped[float] = mapped_column(Numeric(5, 2), nullable=False, default=85)
    utilization_webhook_url: Mapped[str | None] = mapped_column(Text)

    region: Mapped[IPAMRegion] = relationship(back_populates="subnets")
    addresses: Mapped[list["IPAMAddress"]] = relationship(back_populates="subnet", cascade="all, delete-orphan")


class IPAMAddress(Base):
    __tablename__ = "ipam_addresses"
    __table_args__ = (UniqueConstraint("subnet_id", "ip_address", name="unique_ip_per_subnet"),)

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    subnet_id: Mapped[int] = mapped_column(ForeignKey("ipam_subnets.id", ondelete="CASCADE"), nullable=False)
    ip_address: Mapped[str] = mapped_column(INET, nullable=False)
    status: Mapped[IPAMAddressStatus] = mapped_column(ipam_status_pg, nullable=False, default=IPAMAddressStatus.FREE)
    hostname: Mapped[str | None] = mapped_column(String(255))
    mac_address: Mapped[str | None] = mapped_column(MACADDR)
    description: Mapped[str | None] = mapped_column(Text)
    last_seen: Mapped[object | None] = mapped_column(DateTime(timezone=True))
    is_discovered_by_nmap: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    expires_at: Mapped[object | None] = mapped_column(DateTime(timezone=True))
    noc_device_id: Mapped[str | None] = mapped_column(UUID(as_uuid=False))
    dhcp_lease_expires: Mapped[object | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[object] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[object] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    subnet: Mapped[IPAMSubnet] = relationship(back_populates="addresses")


class IPAMAuditLog(Base):
    __tablename__ = "ipam_audit_log"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    entity_type: Mapped[str] = mapped_column(String(32), nullable=False)
    entity_id: Mapped[str] = mapped_column(String(64), nullable=False)
    action: Mapped[str] = mapped_column(String(32), nullable=False)
    actor: Mapped[str | None] = mapped_column(String(255))
    changes: Mapped[dict | None] = mapped_column(JSONB)
    created_at: Mapped[object] = mapped_column(DateTime(timezone=True), server_default=func.now())


class IPAMDHCPLease(Base):
    __tablename__ = "ipam_dhcp_leases"
    __table_args__ = (UniqueConstraint("subnet_id", "ip_address", name="ipam_dhcp_leases_subnet_ip_key"),)

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    subnet_id: Mapped[int | None] = mapped_column(ForeignKey("ipam_subnets.id", ondelete="CASCADE"))
    ip_address: Mapped[str] = mapped_column(INET, nullable=False)
    mac_address: Mapped[str | None] = mapped_column(MACADDR)
    hostname: Mapped[str | None] = mapped_column(String(255))
    expires_at: Mapped[object | None] = mapped_column(DateTime(timezone=True))
    imported_at: Mapped[object] = mapped_column(DateTime(timezone=True), server_default=func.now())
